### ISSUE TITLE
Adapt new transpose kernel that handles small width/height

### ISSUE THREAD

**Issue by RasmusWL** [HIPERFIT/futhark#308]
Currently the transposition kernel used in Futhark is *very* slow if the width or height of the input array is smaller than the tile size (which is 16 right now).

I came up with special transpositions kernel to handle the case of a low width, and a low height. Below is the transposition kernel for handling a low height of the input array. It will use `1/mulx` of its threads in the y-dimension to read a total of `mulx` tiles in the x-dimension of the input array.

```
// Will consume `mulx * TILE` elements of each row of the input array, thereby
// making up for the lack of height in the input array
template <int TILE>
__global__ void transpose_new_lowheight(const float* A, float* B, int heightA, int widthA, int mulx) {

    __shared__ float tile[TILE][TILE+1];

    int x = blockIdx.x * TILE * mulx + threadIdx.x + (threadIdx.y % mulx)*TILE;
    int y = blockIdx.y * TILE + (threadIdx.y / mulx);

    int ind = y * widthA + x;
    if( x < widthA && y < heightA ) {
        tile[threadIdx.y][threadIdx.x] = A[ind];
    }
    __syncthreads();

    x = blockIdx.y * TILE + (threadIdx.x / mulx);
    y = blockIdx.x * TILE * mulx + threadIdx.y + (threadIdx.x % mulx)*TILE;

    ind = y * heightA + x;
    if( x < heightA && y < widthA ) {
        B[ind] = tile[threadIdx.x][threadIdx.y];
    }
}
```

We will launch it like this (only launch if we can fit at least the element of 2 whole tiles -- so it won't work for a `[9][???]` array when using a TILE_DIM of 16).

```
if (height <= TILE_DIM/2 && width > TILE_DIM) {
    int mulx = TILE_DIM / height;
    int newwidth = (width+mulx-1)/mulx;

    dim3 dimGrid ((newwidth+TILE_DIM-1)/TILE_DIM, (height+TILE_DIM-1)/TILE_DIM, 1);
    dim3 dimBlock(TILE_DIM, TILE_DIM, 1);

    transpose_new_lowheight<TILE_DIM> <<<dimGrid, dimBlock>>>(d_idata, d_odata, height, width, mulx);
}
```

The full code can be found at: https://gist.github.com/RasmusWL/3adb4a018a81401785f12bdc866c02b5 which is a benchmark of the different ways of performing the transposition. Results from running on our GTX 780 Ti can be seen below.

An other optimization that would be very powerful is performing a memory copy if the width or height of the original is 1 (because in this case a transposition and a memory copy is just the same)

I would be very happy if you @Athas could look at porting this to our OpenCL code generation.

---

```
./transpose-cost 26
================================================================================
Device : GeForce GTX 780 Ti
      prob size      Futhark32      Futhark16    NVIDIA 32x4         new 32         new 16
       2^0 2^26          84026          35494          22761           7367           4224
       2^1 2^25          41384          18056          11691           5359           4220
       2^2 2^24          20962           9421           6172           4511           4118
       2^3 2^23          10680           4980           3704           4229           4066
       2^4 2^22           5580           2811           2673           4187
       2^5 2^21           3145           3122           2400
       2^6 2^20           2707           3186           2378
       2^7 2^19           2728           2779           2378
       2^8 2^18           2779           2711           2444
       2^9 2^17           2748           2734           2422
      2^10 2^16           2756           2740           2392
      2^11 2^15           2753           2773           2373
      2^12 2^14           2768           2907           2371
      2^13 2^13           2803           3463           2557
      2^14 2^12           2833           3825           2641
      2^15 2^11           2857           3713           2646
      2^16 2^10           2818           3163           2503
       2^17 2^9           2860           3072           2461
       2^18 2^8           2840           3143           2443
       2^19 2^7           2744           2980           2413
       2^20 2^6           2749           3178           2389
       2^21 2^5           3468           3166           2412
       2^22 2^4           6628           3109           3696           4753
       2^23 2^3          12984           5430           6140           4503           4174
       2^24 2^2          23165          10369          11223           5396           4077
       2^25 2^1          45141          20516          21473           7594           4151
       2^26 2^0          90049          40431          37631          11591           4149
```

**Comment by Athas** [comment]
This shouldn't be too hard, but I don't fully understand when we should pick one version over the other.  And we should only have two versions, right?  The current one and your new one?  I can't parse the sentence "only launch if we can fit at least the element of 2 whole tiles".  Fit where?  And what is a "tile"?  Is it a TILE_WIDTHxTILE_WIDTH section of some array?

**Comment by RasmusWL** [comment]
We should have 3 versions: The current version, a version to handle input with small height (the one posted above), and a version to handle input with small width (posted at the bottom of this comment).

We should pick using the following algorithm
```
if (height == 1 || width == 1) {
    copy
} if (height <= TILE_DIM/2 && width > TILE_DIM) {
    transpose_lowheight
} else if (width <= TILE_DIM/2 && height > TILE_DIM) {
    transpose_lowwidth
else {
    transpose_current
}
```

If you *really* want to, you could merge the "lowheight" and "lowwidth" ones (but I think would cause some kind of performance penalty)

---

Yes, I used the word "tile" as a `TILE_DIM x TILE_DIM` section of an array.

If we use a TILE_DIM of 4, and have a `[2][8]` array

```
[ [ 0,  1,  2,  3,  4,  5,  6,  7]
, [ 8,  9, 10, 11, 12, 13, 14, 15]
]
```

normally we would use 2 tiles to transpose this (one for the first four element
of both rows, and one for the last four of both rows). If we consider the number
of elements the current would process per tile (`height * TILE_DIM`), we should
only launch the "lowheight" kernel if it has room for all the elements that 2
normal tiles would process.

This means that a `[3][16]` array will not benefit from the "lowheight" kernel
when using TILE_DIM of 4, and we should launch the normal transpose kernel.

It might be possible to create an other kernel that can handles the cases where
`TILE_DIM > height > TILE_DIM/2`, but this kernel does not.

---

```
template <int TILE>
__global__ void transpose_new_lowwidth(const float* A, float* B, int heightA, int widthA, int muly) {

    __shared__ float tile[TILE][TILE+1];

    int x = blockIdx.x * TILE + (threadIdx.x / muly);
    int y = blockIdx.y * TILE * muly + threadIdx.y + (threadIdx.x % muly)*TILE;

    int ind = y * widthA + x;
    if( x < widthA && y < heightA ) {
        tile[threadIdx.y][threadIdx.x] = A[ind];
    }
    __syncthreads();

    x = blockIdx.y * TILE * muly + threadIdx.x + (threadIdx.y % muly)*TILE;
    y = blockIdx.x * TILE + (threadIdx.y / muly);

    ind = y * heightA + x;
    if( x < heightA && y < widthA ) {
        B[ind] = tile[threadIdx.x][threadIdx.y];
    }
}
```

**Comment by Athas** [comment]
If you add support for an extra dimension, I'll incorporate these.  Remember that the code generator has the notion of 'map with transpose', not just 'transpose'.

**Comment by RasmusWL** [comment]
We can apply the same solution as as the current transpose kernel code uses [src/Futhark/CodeGen/OpenCL/Kernels.hs:69](https://github.com/HIPERFIT/futhark/blob/master/src/Futhark/CodeGen/OpenCL/Kernels.hs#L69):

```
    // Adjust the input and output arrays for the third dimension.
    our_array_offset = get_global_id(2) * width * height;
    odata += our_array_offset;
    idata += our_array_offset;
```

The current transpose kernel also takes the arguments `uint input_size, uint output_size` which can be used to transpose an array that does not have size `height * width`. My kernels does not use this currently, but it should be straightforward to add this feature as well.

I did not realize how easy it would have been to add the kernels to Futhark. I know you added the infrastructure to support this, so if you want me to do the grunt work of adding the new kernels you can just push the work you have done so far to a new branch, and I wil take a look at it!

**Comment by RasmusWL** [comment]
So I implemented this in the branch [new-transpose](https://github.com/HIPERFIT/futhark/tree/new-transpose).

I have used some automatic testing on a good deal of matrices, and I have found no differences in results from the current algorithm. I stored the output from both versions in files, and checked that the sha1-sum matches. `sha1sum out-*/* | sort | uniq --check-chars 40`

I have seen very nice improvements for large matrices, comparable to the data posted in the table above. For small matrices there might be a slight performance regressions, like using 21 microseconds instead of 19 (for a [1][1024] matrix).

The code could be better at deciding when to use the two new kernels. Currently if we use a tile size of 16x16, and give it a [2][32] array, it will apply the "low height" kernel with `mulx =8`. This will in essence be the same as transposing a [16][4] array, expect that we pay a much higher price to perform indexing. But it will only be a problem for very small matrices, so I don't know how much of a concern this is?

The code definitely needs some comments to explain what is going on, and I feel like the code could be made much simpler by unifying the calculations somehow (right now it feels a bit copy-paste like),
